File: app/src/main/java/com/forkmang/activity/ForgotPassword.java
package com.forkmang.activity;

import static com.forkmang.helper.Constant.SUCCESS_CODE_Ne;

import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.forkmang.R;
import com.forkmang.helper.Constant;
import com.forkmang.helper.StorePrefrence;
import com.forkmang.helper.Utils;
import com.forkmang.network_call.Api;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.textfield.TextInputEditText;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.GetTokenResult;
import com.google.gson.Gson;
import com.google.gson.JsonObject;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.util.Objects;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ForgotPassword extends AppCompatActivity {
    TextInputEditText etv_newpas, etvcnf_pass;
    Context ctx = ForgotPassword.this;
    StorePrefrence storePrefrence;
    private ProgressDialog dialog;
    Button btn_reset;
    String idToken;
    ProgressBar progressBar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_forgot_password);
        storePrefrence=new StorePrefrence(ctx);
        progressBar = findViewById(R.id.progressBar);
        btn_reset = findViewById(R.id.btn_reset);
        etv_newpas = findViewById(R.id.etv_newpas);
        etvcnf_pass = findViewById(R.id.etvcnf_pass);

        etv_newpas.setText("1234567");
        etvcnf_pass.setText("1234567");



        btn_reset.setOnClickListener(v -> {
           /* final Intent mainIntent = new Intent(ForgotPassword.this, FaceLoginPermission.class);
            startActivity(mainIntent);
            finish();*/

            if(etv_newpas.getText().length() > 0)
            {
                if(etvcnf_pass.getText().length() > 0)
                {
                    if(etv_newpas.getText().toString().equals(etvcnf_pass.getText().toString()))
                    {
                        //call api
                        String contact = storePrefrence.getString(Constant.MOBILE);
                        getToken(contact);
                    }
                    else{
                        Toast.makeText(ctx, Constant.PASSWORD_MATCH, Toast.LENGTH_SHORT).show();
                    }
                }
                else{
                    Toast.makeText(ctx, Constant.CNFPASSWORD_MATCH, Toast.LENGTH_SHORT).show();
                }
            }
            else{
                Toast.makeText(ctx, Constant.PASSWORD, Toast.LENGTH_SHORT).show();
            }

        });

    }

    private void callApi_forgetpassword_valid(String contact) {
        progressBar.setVisibility(View.VISIBLE);
        Api.getInfo().forgot_pass(contact, idToken).
                enqueue(new Callback<JsonObject>() {
                    @Override
                    public void onResponse(Call<JsonObject> call, Response<JsonObject> response) {
                        try{
                            if(response.code() == Constant.SUCCESS_CODE_2)
                            {
                                JSONObject jsonObject = new JSONObject(new Gson().toJson(response.body()));
                                //Log.d("Result", jsonObject.toString());
                                if(jsonObject.getString("status").equalsIgnoreCase(SUCCESS_CODE_Ne))
                                {
                                    progressBar.setVisibility(View.GONE);
                                    Toast.makeText(ctx,jsonObject.getString("message"),Toast.LENGTH_SHORT).show();
                                    storePrefrence.setString(Constant.MOBILE, jsonObject.getJSONObject("data").getString("contact"));
                                    //storePrefrence.setString(Constant.TOKEN_FORGOTPASS, jsonObject.getJSONObject("data").getString("token"));
                                    storePrefrence.setBoolean("keeplogin", false);


                                    if (Utils.isNetworkAvailable(ctx)) {
                                        callApi_resetpassword(jsonObject.getJSONObject("data").getString("contact"),
                                                Objects.requireNonNull(etv_newpas.getText()).toString(),
                                                Objects.requireNonNull(etvcnf_pass.getText()).toString(),
                                                jsonObject.getJSONObject("data").getString("token") );

                                    }
                                    else{
                                        Toast.makeText(ctx, Constant.NETWORKEROORMSG, Toast.LENGTH_SHORT).show();
                                    }


                                    //stopProgress();
                                }
                                else{
                                    //stopProgress();
                                    progressBar.setVisibility(View.GONE);
                                    Toast.makeText(ctx, "Error occur please try again", Toast.LENGTH_LONG).show();
                                }
                            }
                            else if(response.code() == Constant.ERROR_CODE)
                            {
                                JSONObject jsonObject = new JSONObject(response.errorBody().string());
                                if(jsonObject.getString("status").equalsIgnoreCase(String.valueOf(Constant.ERROR_CODE)))
                                {
                                    progressBar.setVisibility(View.GONE);
                                    String error_msg = jsonObject.getString("message") ;
                                    Toast.makeText(ctx,error_msg,Toast.LENGTH_SHORT).show();

                                }
                                else{
                                    progressBar.setVisibility(View.GONE);
                                    Toast.makeText(ctx, "Error occur please try again", Toast.LENGTH_LONG).show();
                                }


                            }
                        }
                        catch (JSONException | IOException ex)
                        {
                            ex.printStackTrace();
                            //stopProgress();
                            progressBar.setVisibility(View.GONE);
                            Toast.makeText(ctx, "Error occur please try again", Toast.LENGTH_LONG).show();
                        }
                    }
                    @Override
                    public void onFailure(Call<JsonObject> call, Throwable t) {
                        Toast.makeText(ctx, "Error occur please try again", Toast.LENGTH_LONG).show();
                        //stopProgress();
                        progressBar.setVisibility(View.GONE);
                    }
                });

    }

    private void callApi_resetpassword(String contact, String password, String cnf_password, String token) {
        progressBar.setVisibility(View.VISIBLE);
        Api.getInfo().reset_pass(contact, password, cnf_password, token).
                enqueue(new Callback<JsonObject>() {
                    @Override
                    public void onResponse(Call<JsonObject> call, Response<JsonObject> response) {
                        try{
                            if(response.code() == Constant.SUCCESS_CODE_2)
                            {
                                JSONObject jsonObject = new JSONObject(new Gson().toJson(response.body()));
                                //Log.d("Result", jsonObject.toString());
                                if(jsonObject.getString("status").equalsIgnoreCase(SUCCESS_CODE_Ne))
                                {
                                    progressBar.setVisibility(View.GONE);
                                    final Intent mainIntent = new Intent(ForgotPassword.this, LoginFormActivity.class);
                                    startActivity(mainIntent);
                                    finish();
                                }
                                else{
                                    //stopProgress();
                                    progressBar.setVisibility(View.GONE);
                                    Toast.makeText(ctx, "Error occur please try again", Toast.LENGTH_LONG).show();
                                }
                            }
                            else if(response.code() == Constant.ERROR_CODE)
                            {
                                JSONObject jsonObject = new JSONObject(response.errorBody().string());
                                if(jsonObject.getString("status").equalsIgnoreCase(String.valueOf(Constant.ERROR_CODE)))
                                {
                                    progressBar.setVisibility(View.GONE);
                                    String error_msg = jsonObject.getString("message") ;
                                    Toast.makeText(ctx,error_msg,Toast.LENGTH_SHORT).show();

                                }
                                else{
                                    progressBar.setVisibility(View.GONE);
                                    Toast.makeText(ctx, "Error occur please try again", Toast.LENGTH_LONG).show();
                                }


                            }
                        }
                        catch (JSONException | IOException ex)
                        {
                            ex.printStackTrace();
                            //stopProgress();
                            progressBar.setVisibility(View.GONE);
                            Toast.makeText(ctx, "Error occur please try again", Toast.LENGTH_LONG).show();
                        }
                    }
                    @Override
                    public void onFailure(Call<JsonObject> call, Throwable t) {
                        Toast.makeText(ctx, "Error occur please try again", Toast.LENGTH_LONG).show();
                        //stopProgress();
                        progressBar.setVisibility(View.GONE);
                    }
                });

    }


    private void getToken(String contact) {
        //progressBar.setVisibility(View.VISIBLE);
        FirebaseUser mUser = FirebaseAuth.getInstance().getCurrentUser();
        mUser.getIdToken(true)
                .addOnCompleteListener(new OnCompleteListener<GetTokenResult>() {
                    public void onComplete(@NonNull Task<GetTokenResult> task) {
                        if (task.isSuccessful()) {
                            idToken = task.getResult().getToken();
                            Log.d("token", idToken);
                            //progressBar.setVisibility(View.GONE);



                            if (Utils.isNetworkAvailable(ctx)) {
                                callApi_forgetpassword_valid(contact);
                            }
                            else{
                                Toast.makeText(ctx, Constant.NETWORKEROORMSG, Toast.LENGTH_SHORT).show();
                            }

                            // Send token to your backend via HTTPS
                            // ...
                        } else {
                            // Handle error -> task.getException();
                            //progressBar.setVisibility(View.GONE);
                        }
                    }
                });
    }

}

File: app/src/main/java/com/forkmang/activity/ForgotPassword.kt
package com.forkmang.activity

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.forkmang.R
import com.forkmang.databinding.ActivityForgotPasswordBinding
import com.forkmang.helper.Constant.CNFPASSWORD_MATCH
import com.forkmang.helper.Constant.ERRORMSG
import com.forkmang.helper.Constant.ERROR_CODE
import com.forkmang.helper.Constant.MOBILE
import com.forkmang.helper.Constant.NETWORKEROORMSG
import com.forkmang.helper.Constant.PASSWORD
import com.forkmang.helper.Constant.PASSWORD_MATCH
import com.forkmang.helper.Constant.SUCCESS_CODE
import com.forkmang.helper.Constant.SUCCESS_CODE_n
import com.forkmang.helper.StorePrefrence
import com.forkmang.helper.Utils
import com.forkmang.helper.showToastMessage
import com.forkmang.network_call.Api
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.gson.Gson
import com.google.gson.JsonObject
import org.json.JSONException
import org.json.JSONObject
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.IOException

class ForgotPassword : AppCompatActivity() {

    private val binding by lazy { ActivityForgotPasswordBinding.inflate(layoutInflater) }
    var ctx: Context = this@ForgotPassword
    private val storePreference by lazy { StorePrefrence(this) }
    var idToken: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_forgot_password)

        /*binding.etvNewpas.setText("1234567")
        binding.etvcnfPass.setText("1234567")*/

        binding.btnReset.setOnClickListener {
            if (binding.etvNewpas.text?.isNotEmpty() == true) {
                if (binding.etvcnfPass.text?.isNotEmpty() == true) {
                    if ((binding.etvNewpas.text.toString() == binding.etvcnfPass.text.toString())) {
                        //call api
                        val contact: String? = storePreference.getString(MOBILE)
                        if (contact != null) {
                            //TODO: change the token method
                            getToken(contact)
                        }
                    } else {
                        showToastMessage(PASSWORD_MATCH)
                    }
                } else {
                    showToastMessage(CNFPASSWORD_MATCH)
                }
            } else {
                showToastMessage(PASSWORD)
            }
        }
    }

    private fun callApiForgetPasswordValid(contact: String) {
        binding.progressBar.visibility = View.VISIBLE
        Api.info.forgotPass(contact, idToken)?.enqueue(object : Callback<JsonObject?> {
            override fun onResponse(
                call: Call<JsonObject?>,
                response: Response<JsonObject?>
            ) {
                try {
                    if (response.code() == SUCCESS_CODE_n) {
                        val jsonObject = JSONObject(Gson().toJson(response.body()))
                        //Log.d("Result", jsonObject.toString());
                        if (jsonObject.getString("status")
                                .equals(SUCCESS_CODE, ignoreCase = true)
                        ) {
                            binding.progressBar.visibility = View.GONE
                            showToastMessage(jsonObject.getString("message"))
                            storePreference.setString(
                                MOBILE,
                                jsonObject.getJSONObject("data").getString("contact")
                            )
                            //storePrefrence.setString(Constant.TOKEN_FORGOTPASS, jsonObject.getJSONObject("data").getString("token"));
                            storePreference.setBoolean("keeplogin", false)
                            if (Utils.isNetworkAvailable(ctx)) {
                                callApiResetPassword(
                                    jsonObject.getJSONObject("data").getString("contact"),
                                    binding.etvNewpas.text.toString(),
                                    binding.etvcnfPass.text.toString(),
                                    jsonObject.getJSONObject("data").getString("token")
                                )
                            } else {
                                showToastMessage(NETWORKEROORMSG)
                            }
                        } else {
                            binding.progressBar.visibility = View.GONE
                            showToastMessage(ERRORMSG)
                        }
                    } else if (response.code() == ERROR_CODE) {
                        val jsonObject = JSONObject(response.errorBody()!!.string())
                        if (jsonObject.getString("status")
                                .equals(ERROR_CODE.toString(), ignoreCase = true)
                        ) {
                            binding.progressBar.visibility = View.GONE
                            val errorMsg: String = jsonObject.getString("message")
                            showToastMessage(errorMsg)
                        } else {
                            binding.progressBar.visibility = View.GONE
                            showToastMessage(ERRORMSG)
                        }
                    }
                } catch (ex: JSONException) {
                    ex.printStackTrace()
                    binding.progressBar.visibility = View.GONE
                    showToastMessage(ERRORMSG)
                } catch (ex: IOException) {
                    ex.printStackTrace()
                    binding.progressBar.visibility = View.GONE
                    showToastMessage(ERRORMSG)
                }
            }

            override fun onFailure(call: Call<JsonObject?>, t: Throwable) {
                showToastMessage(ERRORMSG)
                binding.progressBar.visibility = View.GONE
            }
        })
    }

    private fun callApiResetPassword(
        contact: String,
        password: String,
        cnf_password: String,
        token: String
    ) {
        binding.progressBar.visibility = View.VISIBLE
        Api.info.resetPass(contact, password, cnf_password, token)
            ?.enqueue(object : Callback<JsonObject?> {
                override fun onResponse(
                    call: Call<JsonObject?>,
                    response: Response<JsonObject?>
                ) {
                    try {
                        if (response.code() == SUCCESS_CODE_n) {
                            val jsonObject = JSONObject(Gson().toJson(response.body()))
                            //Log.d("Result", jsonObject.toString());
                            if (jsonObject.getString("status")
                                    .equals(SUCCESS_CODE, ignoreCase = true)
                            ) {
                                binding.progressBar.visibility = View.GONE
                                val mainIntent =
                                    Intent(this@ForgotPassword, LoginFormActivity::class.java)
                                startActivity(mainIntent)
                                finish()
                            } else {
                                binding.progressBar.visibility = View.GONE
                                showToastMessage(ERRORMSG)
                            }
                        } else if (response.code() == ERROR_CODE) {
                            val jsonObject = JSONObject(response.errorBody()!!.string())
                            if (jsonObject.getString("status")
                                    .equals(ERROR_CODE.toString(), ignoreCase = true)
                            ) {
                                binding.progressBar.visibility = View.GONE
                                val errorMsg: String = jsonObject.getString("message")
                                showToastMessage(errorMsg)
                            } else {
                                binding.progressBar.visibility = View.GONE
                                showToastMessage(ERRORMSG)
                            }
                        }
                    } catch (ex: JSONException) {
                        ex.printStackTrace()
                        binding.progressBar.visibility = View.GONE
                        showToastMessage(ERRORMSG)
                    } catch (ex: IOException) {
                        ex.printStackTrace()
                        binding.progressBar.visibility = View.GONE
                        showToastMessage(ERRORMSG)
                    }
                }

                override fun onFailure(call: Call<JsonObject?>, t: Throwable) {
                    showToastMessage(ERRORMSG)
                    binding.progressBar.visibility = View.GONE
                }
            })
    }

    private fun getToken(contact: String) {
        //progressBar.setVisibility(View.VISIBLE);
        val mUser: FirebaseUser? = FirebaseAuth.getInstance().currentUser
        mUser?.getIdToken(true)
            ?.addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    idToken = task.result.token
                    Log.d("token", (idToken) ?: "null")
                    //progressBar.setVisibility(View.GONE);
                    if (Utils.isNetworkAvailable(ctx)) {
                        callApiForgetPasswordValid(contact)
                    } else {
                        showToastMessage(NETWORKEROORMSG)
                    }

                    // Send token to your backend via HTTPS
                    // ...
                } else {
                    // Handle error -> task.getException();
                    //progressBar.setVisibility(View.GONE);
                }
            }
    }
}